build: header-only Boost

fanquake opened the pull request:
This PR converts our Boost usage to header only. We switch from using our last remaining Boost lib (unit test), to using it's header-only implementation (see https://www.boost.org/doc/libs/1_78_0/libs/test/doc/html/boost_test/adv_scenarios/single_header_customizations/multiple_translation_units.html).

Also related to #24291.

Guix build:
```bash
```

fanquake committed:
build: use header-only Boost unit test

fanquake committed:
build: header-only Boost

fanquake committed:
build: remove native B2 package

MarcoFalke commented:
Didn't know that boost unit test is also header-only.

Concept ACK, I guess :rocket:

hebasto commented:
Concept ACK.

laanwj commented:
Concept ACK, this is a nice cleanup to configure.ac.

laanwj commented on depends/packages/boost.mk:
```diff
@@ -3,45 +3,8 @@ $(package)_version=1_71_0
 define $(package)_stage_cmds
-  b2 -d0 -j4 --prefix=$($(package)_staging_prefix_dir) $($(package)_config_opts) toolset=$($(package)_toolset_$(host_os)) --no-cmake-config install
+  mkdir -p $($(package)_staging_prefix_dir)/include && \
+  cp -r boost $($(package)_staging_prefix_dir)/include
```
Nice.

laanwj commented on configure.ac:
```diff
@@ -1476,30 +1476,6 @@ if test "$use_tests" = "yes"; then
-    LIBS="$TEMP_LIBS"
-    CPPFLAGS="$TEMP_CPPFLAGS"
-
-  fi
```
`build-aux/m4/ax_boost_unit_test_framework.m4` can go after this?

fanquake replied:
Yes; will be dropped in the next push.

fanquake commented:
Updated to fix the linter, added a commit to use Boost 1.78.0, and removed the Boost Unit Test macro.

> Does this bump the minimum req. boost version?

I'll need to double check this.

TSAN failure from previous change:
```bash
==================
WARNING: ThreadSanitizer: data race (pid=90566)
  Read of size 4 at 0x7fcbd5a62f40 by thread T4 (mutexes: write M1909):
    #0 std::__1::ios_base::flags() const /usr/lib/llvm-13/bin/../include/c++/v1/ios:461:12 (test_bitcoin+0x2192b8)
    #1 std::__1::basic_ostream<char, std::__1::char_traits<char> >& std::__1::__put_character_sequence<char, std::__1::char_traits<char> >(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, char const*, unsigned long) /usr/lib/llvm-13/bin/../include/c++/v1/ostream:720:40 (test_bitcoin+0x2192b8)
    #2 std::__1::basic_ostream<char, std::__1::char_traits<char> >& std::__1::operator<<<char, std::__1::char_traits<char>, std::__1::allocator<char> >(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) /usr/lib/llvm-13/bin/../include/c++/v1/ostream:1041:12 (test_bitcoin+0x1d94d7)
    #3 $_0::operator()(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const src/test/main.cpp:28:15 (test_bitcoin+0x1d94d7)
    #4 decltype(static_cast<$_0&>(fp)(static_cast<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&>(fp0))) std::__1::__invoke<$_0&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&>($_0&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) /usr/lib/llvm-13/bin/../include/c++/v1/type_traits:3918:1 (test_bitcoin+0x1d94d7)
    #5 void std::__1::__invoke_void_return_wrapper<void, true>::__call<$_0&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&>($_0&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) /usr/lib/llvm-13/bin/../include/c++/v1/__functional/invoke.h:61:9 (test_bitcoin+0x1d94d7)
    #6 std::__1::__function::__alloc_func<$_0, std::__1::allocator<$_0>, void (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)>::operator()(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) /usr/lib/llvm-13/bin/../include/c++/v1/__functional/function.h:171:16 (test_bitcoin+0x1d94d7)
    #7 std::__1::__function::__func<$_0, std::__1::allocator<$_0>, void (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)>::operator()(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) /usr/lib/llvm-13/bin/../include/c++/v1/__functional/function.h:345:12 (test_bitcoin+0x1d94d7)
    #8 std::__1::__function::__value_func<void (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)>::operator()(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const /usr/lib/llvm-13/bin/../include/c++/v1/__functional/function.h:498:16 (test_bitcoin+0xe8104b)
    #9 std::__1::function<void (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)>::operator()(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const /usr/lib/llvm-13/bin/../include/c++/v1/__functional/function.h:1175:12 (test_bitcoin+0xe8104b)
    #10 BCLog::Logger::LogPrintStr(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) src/logging.cpp:279:9 (test_bitcoin+0xe8104b)
    #11 void LogPrintf_<char const*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, char const*, char const* const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) src/./logging.h:184:23 (test_bitcoin+0xd7ec50)
    #12 CMainSignals::TransactionRemovedFromMempool(std::__1::shared_ptr<CTransaction const> const&, MemPoolRemovalReason, unsigned long)::$_7::operator()() const src/validationinterface.cpp:215:5 (test_bitcoin+0xd84022)
    #13 decltype(static_cast<CMainSignals::TransactionRemovedFromMempool(std::__1::shared_ptr<CTransaction const> const&, MemPoolRemovalReason, unsigned long)::$_7&>(fp)()) std::__1::__invoke<CMainSignals::TransactionRemovedFromMempool(std::__1::shared_ptr<CTransaction const> const&, MemPoolRemovalReason, unsigned long)::$_7&>(CMainSignals::TransactionRemovedFromMempool(std::__1::shared_ptr<CTransaction const> const&, MemPoolRemovalReason, unsigned long)::$_7&) /usr/lib/llvm-13/bin/../include/c++/v1/type_traits:3918:1 (test_bitcoin+0xd84022)
    #14 void std::__1::__invoke_void_return_wrapper<void, true>::__call<CMainSignals::TransactionRemovedFromMempool(std::__1::shared_ptr<CTransaction const> const&, MemPoolRemovalReason, unsigned long)::$_7&>(CMainSignals::TransactionRemovedFromMempool(std::__1::shared_ptr<CTransaction const> const&, MemPoolRemovalReason, unsigned long)::$_7&) /usr/lib/llvm-13/bin/../include/c++/v1/__functional/invoke.h:61:9 (test_bitcoin+0xd84022)
    #15 std::__1::__function::__alloc_func<CMainSignals::TransactionRemovedFromMempool(std::__1::shared_ptr<CTransaction const> const&, MemPoolRemovalReason, unsigned long)::$_7, std::__1::allocator<CMainSignals::TransactionRemovedFromMempool(std::__1::shared_ptr<CTransaction const> const&, MemPoolRemovalReason, unsigned long)::$_7>, void ()>::operator()() /usr/lib/llvm-13/bin/../include/c++/v1/__functional/function.h:171:16 (test_bitcoin+0xd84022)
    #16 std::__1::__function::__func<CMainSignals::TransactionRemovedFromMempool(std::__1::shared_ptr<CTransaction const> const&, MemPoolRemovalReason, unsigned long)::$_7, std::__1::allocator<CMainSignals::TransactionRemovedFromMempool(std::__1::shared_ptr<CTransaction const> const&, MemPoolRemovalReason, unsigned long)::$_7>, void ()>::operator()() /usr/lib/llvm-13/bin/../include/c++/v1/__functional/function.h:345:12 (test_bitcoin+0xd84022)
    #17 std::__1::__function::__value_func<void ()>::operator()() const /usr/lib/llvm-13/bin/../include/c++/v1/__functional/function.h:498:16 (test_bitcoin+0xe2a2b1)
    #18 std::__1::function<void ()>::operator()() const /usr/lib/llvm-13/bin/../include/c++/v1/__functional/function.h:1175:12 (test_bitcoin+0xe2a2b1)
    #19 SingleThreadedSchedulerClient::ProcessQueue() src/scheduler.cpp:177:5 (test_bitcoin+0xe2a2b1)
    #20 decltype(*(static_cast<SingleThreadedSchedulerClient*&>(fp0)).*fp()) std::__1::__invoke<void (SingleThreadedSchedulerClient::*&)(), SingleThreadedSchedulerClient*&, void>(void (SingleThreadedSchedulerClient::*&)(), SingleThreadedSchedulerClient*&) /usr/lib/llvm-13/bin/../include/c++/v1/type_traits:3859:1 (test_bitcoin+0xe2c7ea)
    #21 std::__1::__bind_return<void (SingleThreadedSchedulerClient::*)(), std::__1::tuple<SingleThreadedSchedulerClient*>, std::__1::tuple<>, __is_valid_bind_return<void (SingleThreadedSchedulerClient::*)(), std::__1::tuple<SingleThreadedSchedulerClient*>, std::__1::tuple<> >::value>::type std::__1::__apply_functor<void (SingleThreadedSchedulerClient::*)(), std::__1::tuple<SingleThreadedSchedulerClient*>, 0ul, std::__1::tuple<> >(void (SingleThreadedSchedulerClient::*&)(), std::__1::tuple<SingleThreadedSchedulerClient*>&, std::__1::__tuple_indices<0ul>, std::__1::tuple<>&&) /usr/lib/llvm-13/bin/../include/c++/v1/__functional/bind.h:257:12 (test_bitcoin+0xe2c7ea)
    #22 std::__1::__bind_return<void (SingleThreadedSchedulerClient::*)(), std::__1::tuple<SingleThreadedSchedulerClient*>, std::__1::tuple<>, __is_valid_bind_return<void (SingleThreadedSchedulerClient::*)(), std::__1::tuple<SingleThreadedSchedulerClient*>, std::__1::tuple<> >::value>::type std::__1::__bind<void (SingleThreadedSchedulerClient::*)(), SingleThreadedSchedulerClient*>::operator()<>() /usr/lib/llvm-13/bin/../include/c++/v1/__functional/bind.h:292:20 (test_bitcoin+0xe2c7ea)
    #23 decltype(static_cast<std::__1::__bind<void (SingleThreadedSchedulerClient::*)(), SingleThreadedSchedulerClient*>&>(fp)()) std::__1::__invoke<std::__1::__bind<void (SingleThreadedSchedulerClient::*)(), SingleThreadedSchedulerClient*>&>(std::__1::__bind<void (SingleThreadedSchedulerClient::*)(), SingleThreadedSchedulerClient*>&) /usr/lib/llvm-13/bin/../include/c++/v1/type_traits:3918:1 (test_bitcoin+0xe2c7ea)
    #24 void std::__1::__invoke_void_return_wrapper<void, true>::__call<std::__1::__bind<void (SingleThreadedSchedulerClient::*)(), SingleThreadedSchedulerClient*>&>(std::__1::__bind<void (SingleThreadedSchedulerClient::*)(), SingleThreadedSchedulerClient*>&) /usr/lib/llvm-13/bin/../include/c++/v1/__functional/invoke.h:61:9 (test_bitcoin+0xe2c7ea)
    #25 std::__1::__function::__alloc_func<std::__1::__bind<void (SingleThreadedSchedulerClient::*)(), SingleThreadedSchedulerClient*>, std::__1::allocator<std::__1::__bind<void (SingleThreadedSchedulerClient::*)(), SingleThreadedSchedulerClient*> >, void ()>::operator()() /usr/lib/llvm-13/bin/../include/c++/v1/__functional/function.h:171:16 (test_bitcoin+0xe2c7ea)
    #26 std::__1::__function::__func<std::__1::__bind<void (SingleThreadedSchedulerClient::*)(), SingleThreadedSchedulerClient*>, std::__1::allocator<std::__1::__bind<void (SingleThreadedSchedulerClient::*)(), SingleThreadedSchedulerClient*> >, void ()>::operator()() /usr/lib/llvm-13/bin/../include/c++/v1/__functional/function.h:345:12 (test_bitcoin+0xe2c7ea)
    #27 std::__1::__function::__value_func<void ()>::operator()() const /usr/lib/llvm-13/bin/../include/c++/v1/__functional/function.h:498:16 (test_bitcoin+0xe292bc)
    #28 std::__1::function<void ()>::operator()() const /usr/lib/llvm-13/bin/../include/c++/v1/__functional/function.h:1175:12 (test_bitcoin+0xe292bc)
    #29 CScheduler::serviceQueue() src/scheduler.cpp:64:17 (test_bitcoin+0xe292bc)
    #30 ChainTestingSetup::ChainTestingSetup(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::vector<char const*, std::__1::allocator<char const*> > const&)::$_0::operator()() const src/test/util/setup_common.cpp:156:110 (test_bitcoin+0x9bdaa8)
    #31 decltype(static_cast<ChainTestingSetup::ChainTestingSetup(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::vector<char const*, std::__1::allocator<char const*> > const&)::$_0&>(fp)()) std::__1::__invoke<ChainTestingSetup::ChainTestingSetup(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::vector<char const*, std::__1::allocator<char const*> > const&)::$_0&>(ChainTestingSetup::ChainTestingSetup(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::vector<char const*, std::__1::allocator<char const*> > const&)::$_0&) /usr/lib/llvm-13/bin/../include/c++/v1/type_traits:3918:1 (test_bitcoin+0x9bdaa8)
    #32 void std::__1::__invoke_void_return_wrapper<void, true>::__call<ChainTestingSetup::ChainTestingSetup(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::vector<char const*, std::__1::allocator<char const*> > const&)::$_0&>(ChainTestingSetup::ChainTestingSetup(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::vector<char const*, std::__1::allocator<char const*> > const&)::$_0&) /usr/lib/llvm-13/bin/../include/c++/v1/__functional/invoke.h:61:9 (test_bitcoin+0x9bdaa8)
    #33 std::__1::__function::__alloc_func<ChainTestingSetup::ChainTestingSetup(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::vector<char const*, std::__1::allocator<char const*> > const&)::$_0, std::__1::allocator<ChainTestingSetup::ChainTestingSetup(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::vector<char const*, std::__1::allocator<char const*> > const&)::$_0>, void ()>::operator()() /usr/lib/llvm-13/bin/../include/c++/v1/__functional/function.h:171:16 (test_bitcoin+0x9bdaa8)
    #34 std::__1::__function::__func<ChainTestingSetup::ChainTestingSetup(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::vector<char const*, std::__1::allocator<char const*> > const&)::$_0, std::__1::allocator<ChainTestingSetup::ChainTestingSetup(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::vector<char const*, std::__1::allocator<char const*> > const&)::$_0>, void ()>::operator()() /usr/lib/llvm-13/bin/../include/c++/v1/__functional/function.h:345:12 (test_bitcoin+0x9bdaa8)
    #35 std::__1::__function::__value_func<void ()>::operator()() const /usr/lib/llvm-13/bin/../include/c++/v1/__functional/function.h:498:16 (test_bitcoin+0xebff6f)
    #36 std::__1::function<void ()>::operator()() const /usr/lib/llvm-13/bin/../include/c++/v1/__functional/function.h:1175:12 (test_bitcoin+0xebff6f)
    #37 util::TraceThread(char const*, std::__1::function<void ()>) src/util/thread.cpp:18:9 (test_bitcoin+0xebff6f)
    #38 decltype(static_cast<void (*>(fp)(static_cast<char const*>(fp0), static_cast<ChainTestingSetup::ChainTestingSetup(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::vector<char const*, std::__1::allocator<char const*> > const&)::$_0>(fp0))) std::__1::__invoke<void (*)(char const*, std::__1::function<void ()>), char const*, ChainTestingSetup::ChainTestingSetup(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::vector<char const*, std::__1::allocator<char const*> > const&)::$_0>(void (*&&)(char const*, std::__1::function<void ()>), char const*&&, ChainTestingSetup::ChainTestingSetup(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::vector<char const*, std::__1::allocator<char const*> > const&)::$_0&&) /usr/lib/llvm-13/bin/../include/c++/v1/type_traits:3918:1 (test_bitcoin+0x9bd6a1)
    #39 void std::__1::__thread_execute<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void (*)(char const*, std::__1::function<void ()>), char const*, ChainTestingSetup::ChainTestingSetup(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::vector<char const*, std::__1::allocator<char const*> > const&)::$_0, 2ul, 3ul>(std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void (*)(char const*, std::__1::function<void ()>), char const*, ChainTestingSetup::ChainTestingSetup(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::vector<char const*, std::__1::allocator<char const*> > const&)::$_0>&, std::__1::__tuple_indices<2ul, 3ul>) /usr/lib/llvm-13/bin/../include/c++/v1/thread:280:5 (test_bitcoin+0x9bd6a1)
    #40 void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void (*)(char const*, std::__1::function<void ()>), char const*, ChainTestingSetup::ChainTestingSetup(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::vector<char const*, std::__1::allocator<char const*> > const&)::$_0> >(void*) /usr/lib/llvm-13/bin/../include/c++/v1/thread:291:5 (test_bitcoin+0x9bd6a1)
```

theuni commented:
Nice! Concept ACK.

MarcoFalke commented:
> TSAN failure from previous change:

I haven't checked in detail, but this looks like a known bug. Maybe adjust the suppressions for now?

fanquake commented:
Added a suppression, removed a suppression, pruned out `BOOST_LIBS`.

fanquake commented:
Removed the Boost bump (no need to do that here, and it may have other issues), and have marked this as ready for review. The failure in the ARM CI looks unrelated. Updated the PR description as well.

fanquake commented:
> So I guess this doesn't change the minimum required.

Thanks for taking a look.

Guix build added to PR description.

theStack commented:
Concept ACK

sipa commented:
Concept ACK!

fanquake commented:
Rebased, and added some more changes to the first commit. We don't need to install `libboost-test-dev ` any more, so I've removed that from the docs and CI.

hebasto commented:
Is `depends/patches/boost/fix_openbsd_test_lib.patch` required for headers-only version of Boost.Test?

(cannot check it by myself)

hebasto commented on src/test/main.cpp:
```diff
@@ -3,11 +3,13 @@
 
 /**
- * See https://www.boost.org/doc/libs/1_71_0/libs/test/doc/html/boost_test/utf_reference/link_references/link_boost_test_module_macro.html
+ * See https://www.boost.org/doc/libs/1_78_0/libs/test/doc/html/boost_test/adv_scenarios/single_header_customizations/multiple_translation_units.html
```
Considering that our minimum required boost version is 1.64.0, maybe point at the earliest available version of the same page:
```suggestion
 * See https://www.boost.org/doc/libs/1_69_0/libs/test/doc/html/boost_test/adv_scenarios/single_header_customizations/multiple_translation_units.html
```
?

hebasto reviewed (commented):
Approach ACK 571ae1b68b9e6bccfea35859ee2959ef65876083.

From my understanding of https://www.boost.org/doc/libs/1_68_0/libs/test/doc/adv_scenarios/single_header_customizations.qbk, some changes are required for a single translation unit, therefore, they could be reverted back:
```diff
--- a/src/test/main.cpp
+++ b/src/test/main.cpp
@@ -6,8 +6,6 @@
  * See https://www.boost.org/doc/libs/1_78_0/libs/test/doc/html/boost_test/adv_scenarios/single_header_customizations/multiple_translation_units.html
  */
 #define BOOST_TEST_MODULE Bitcoin Core Test Suite
-#define BOOST_TEST_NO_MAIN
-#define BOOST_TEST_ALTERNATIVE_INIT_API
 
 #include <boost/test/included/unit_test.hpp>
 
@@ -41,8 +39,3 @@ const std::function<std::vector<const char*>()> G_TEST_COMMAND_LINE_ARGUMENTS =
     }
     return args;
 };
-
-int main(int argc, char* argv[], char* envp[])
-{
-  return boost::unit_test::unit_test_main(init_unit_test, argc, argv);
-}
```

hebasto commented:
#### Guix builds:
```
$ find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
e66d35475e19f6cfef6ee79c6d5ec2d214b0bed4ece32093fe409f94c10c23eb  guix-build-571ae1b68b9e/output/aarch64-linux-gnu/SHA256SUMS.part
341c236e55af9102d443363f282cfef37998472b5cc108008eb6e229b1eaa37d  guix-build-571ae1b68b9e/output/aarch64-linux-gnu/bitcoin-571ae1b68b9e-aarch64-linux-gnu-debug.tar.gz
9b4c803c897aa442e2817592b4d2aa707a671663abd989bb8e77751b2736bb09  guix-build-571ae1b68b9e/output/aarch64-linux-gnu/bitcoin-571ae1b68b9e-aarch64-linux-gnu.tar.gz
92e2143236bcb160f3f271b8c892bc853c50bf05e1fa1b396dc708a36be82dac  guix-build-571ae1b68b9e/output/arm-linux-gnueabihf/SHA256SUMS.part
7535e69488cff3091d6bec05b4644239799638a8692b6b15a2329884b2704e6b  guix-build-571ae1b68b9e/output/arm-linux-gnueabihf/bitcoin-571ae1b68b9e-arm-linux-gnueabihf-debug.tar.gz
e30fbde2c459015dda328a1fd8098fac7d7f5759ebeb9a062c4ff77fea3a6ebf  guix-build-571ae1b68b9e/output/arm-linux-gnueabihf/bitcoin-571ae1b68b9e-arm-linux-gnueabihf.tar.gz
45934667cb52591437b27e21f6ae6fa5995a79aaf47b023439611bfbefe04c61  guix-build-571ae1b68b9e/output/arm64-apple-darwin/SHA256SUMS.part
3f1886febc5304181e5de792cf3c9606779c3fca7dc5b921db772746c03211bd  guix-build-571ae1b68b9e/output/arm64-apple-darwin/bitcoin-571ae1b68b9e-arm64-apple-darwin.tar.gz
74c0453f2ff6a605d7c31b58324bb863c2976ab908d5f1f7098370f858a9ce16  guix-build-571ae1b68b9e/output/arm64-apple-darwin/bitcoin-571ae1b68b9e-osx-unsigned.dmg
81c1ec6a691a333e12268fe8d24e3e7b3fdcc30e9b40cccdf7fa602e2ff2ce1e  guix-build-571ae1b68b9e/output/arm64-apple-darwin/bitcoin-571ae1b68b9e-osx-unsigned.tar.gz
1ff2e5f5e3b04aea3edb3a3a36fffc85bf3d8d4039c277d1412dd5b46461d275  guix-build-571ae1b68b9e/output/dist-archive/bitcoin-571ae1b68b9e.tar.gz
600fbcb6285964b098a67c7b737a80b5607b39ef8ce5d38b741374b854a02b41  guix-build-571ae1b68b9e/output/powerpc64-linux-gnu/SHA256SUMS.part
b78e5b09f87f1122d2f448dfa94e6a2a1b8f46c27c77e0d2c3bd95a63706bba9  guix-build-571ae1b68b9e/output/powerpc64-linux-gnu/bitcoin-571ae1b68b9e-powerpc64-linux-gnu-debug.tar.gz
2aca6c47f5a350ab4857b1a3ef80e6be00bcd874588436b6177c75ca25e0f994  guix-build-571ae1b68b9e/output/powerpc64-linux-gnu/bitcoin-571ae1b68b9e-powerpc64-linux-gnu.tar.gz
f953b5cc55cb0b61496468e090a8cd52118083174ebc02881c0900d8282918cc  guix-build-571ae1b68b9e/output/powerpc64le-linux-gnu/SHA256SUMS.part
8d220214ac7135d05a69f27bc38ef508b39674113fdd89f52cef0840fb04a9dd  guix-build-571ae1b68b9e/output/powerpc64le-linux-gnu/bitcoin-571ae1b68b9e-powerpc64le-linux-gnu-debug.tar.gz
620c6247866526f1937f5e9761daf4cbbc58f649b523b056e5d01c02c1592bdf  guix-build-571ae1b68b9e/output/powerpc64le-linux-gnu/bitcoin-571ae1b68b9e-powerpc64le-linux-gnu.tar.gz
aa6fa6f64fffd50ab166753049c961c1bd57a11d66d337bfccc5865df2431858  guix-build-571ae1b68b9e/output/riscv64-linux-gnu/SHA256SUMS.part
3b6533546d5f08ab0b37bdf42fac9a58acf8ec2f14aad31ceda4480eaa0e2408  guix-build-571ae1b68b9e/output/riscv64-linux-gnu/bitcoin-571ae1b68b9e-riscv64-linux-gnu-debug.tar.gz
8f2371f1693fd85f1815295be76f91983f2e734c463dbad567982b907207221d  guix-build-571ae1b68b9e/output/riscv64-linux-gnu/bitcoin-571ae1b68b9e-riscv64-linux-gnu.tar.gz
03107719387f739208c3a16d915a09367b68d9d845a1f2dc1497f586ee8edc7d  guix-build-571ae1b68b9e/output/x86_64-apple-darwin/SHA256SUMS.part
904b2dc8119839713560279312473f15b5af0b4e0ef30f550ec959942854ae25  guix-build-571ae1b68b9e/output/x86_64-apple-darwin/bitcoin-571ae1b68b9e-osx-unsigned.dmg
eff39df332d85ef73cd50cec1f6b2b9dfa3fb1ae7110e533deb9de8ec519e58f  guix-build-571ae1b68b9e/output/x86_64-apple-darwin/bitcoin-571ae1b68b9e-osx-unsigned.tar.gz
768d687ac5d92132e629c990585b97a03b135215346f5457728456f4f3f67894  guix-build-571ae1b68b9e/output/x86_64-apple-darwin/bitcoin-571ae1b68b9e-osx64.tar.gz
55fdb506bfa073da64dfbfc5fe61bd359afce11adf304cae2461a8d37bd91d7f  guix-build-571ae1b68b9e/output/x86_64-linux-gnu/SHA256SUMS.part
d97635485b3e5644d9731cf4d4ed146161410ed590faf160dcae08ae1d84a08d  guix-build-571ae1b68b9e/output/x86_64-linux-gnu/bitcoin-571ae1b68b9e-x86_64-linux-gnu-debug.tar.gz
3caf5d1c6e39a77be52c0a384094c335b8a8a6a99d9e0615ea6aeb08c0c41646  guix-build-571ae1b68b9e/output/x86_64-linux-gnu/bitcoin-571ae1b68b9e-x86_64-linux-gnu.tar.gz
```

fanquake commented on depends/packages/boost.mk:
```diff
@@ -3,48 +3,8 @@ $(package)_version=1.71.0
 define $(package)_stage_cmds
-  b2 -d0 -j4 --prefix=$($(package)_staging_prefix_dir) $($(package)_config_opts) toolset=$($(package)_toolset_$(host_os)) --no-cmake-config install
+  mkdir -p $($(package)_staging_prefix_dir)/include && \
+  cp -r boost $($(package)_staging_prefix_dir)/include
```
This isn't going to work as-is, because the headers are not self-contained enough, and will include headers from other parts of Boost. It's also missing version.hpp, which would be needed for the configure checks, and config.hpp. I think a better way to do this could be to use something like bcp, but that can be done in a follow up.

hebasto reviewed (approved):
ACK 46b905163530e86ea86a89b0950a88ad4aa03761

hebasto commented:
#### Guix builds:
```
$ find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
696f9eeedc8436dfaa6bf4376a00950ae870c4a1df150430e3cabbcbe270761a  guix-build-46b905163530/output/aarch64-linux-gnu/SHA256SUMS.part
942903711be8139face084758ed597f0658a54f0976a6de2387e4a0b3e48c321  guix-build-46b905163530/output/aarch64-linux-gnu/bitcoin-46b905163530-aarch64-linux-gnu-debug.tar.gz
1d33af2b9eb8b90a0ed133acdb2a862d4d5fc972b09e624f10b8363c4de7b679  guix-build-46b905163530/output/aarch64-linux-gnu/bitcoin-46b905163530-aarch64-linux-gnu.tar.gz
fe1dcae66447348c647e3777fd1324fffc2c7276684b2a12e86a8a33bec5525e  guix-build-46b905163530/output/arm-linux-gnueabihf/SHA256SUMS.part
344ec0ea8b94dab1e12786c60bff75069ba0eedce6b292c8b1f55ac54571b819  guix-build-46b905163530/output/arm-linux-gnueabihf/bitcoin-46b905163530-arm-linux-gnueabihf-debug.tar.gz
2c9d0238e1a31155a6cefe07c7ae92fcf4897d6dc952169024d56a192a2d8e7e  guix-build-46b905163530/output/arm-linux-gnueabihf/bitcoin-46b905163530-arm-linux-gnueabihf.tar.gz
0e2ab4836244c63295ac390b718e8163d0323c48e6a59fbf0160b136c15cca3c  guix-build-46b905163530/output/arm64-apple-darwin/SHA256SUMS.part
da224a7c6acfe387a970d43835aca8e108d3ae131f2b5e78d6635eafdf3ba64d  guix-build-46b905163530/output/arm64-apple-darwin/bitcoin-46b905163530-arm64-apple-darwin.tar.gz
e774bf43d02567d30b8a91c4e5b267428601f86c10c84b14a78ecc4060fe277c  guix-build-46b905163530/output/arm64-apple-darwin/bitcoin-46b905163530-osx-unsigned.dmg
0633977d7c9655af3ff96ce1afec49deae80669358a60b4edbc6c121a68d34c1  guix-build-46b905163530/output/arm64-apple-darwin/bitcoin-46b905163530-osx-unsigned.tar.gz
870c3c6a12622611ce4c556eb82d192f6f8ff91017053f3ded3e6def1d364ddb  guix-build-46b905163530/output/dist-archive/bitcoin-46b905163530.tar.gz
96f7c916883f95a331107081aa76e293bf5f74b868ca548dcd7dbbeadb37fff2  guix-build-46b905163530/output/powerpc64-linux-gnu/SHA256SUMS.part
d9f93fb9d01335446a838cb4154989404b5da43b5e2b30ede38282e09cf1c4bf  guix-build-46b905163530/output/powerpc64-linux-gnu/bitcoin-46b905163530-powerpc64-linux-gnu-debug.tar.gz
77b345e81fdada5b0a38b609bcee9713caf52e217630a52417c04842fec5f76f  guix-build-46b905163530/output/powerpc64-linux-gnu/bitcoin-46b905163530-powerpc64-linux-gnu.tar.gz
56db166ee61b414615eb44206d4c2345f5b80205f51928157d15e529ac5d6817  guix-build-46b905163530/output/powerpc64le-linux-gnu/SHA256SUMS.part
3dd0b6487a93ff14c682cdcd34a712f7cca5320887222f16a251e046f3ec0342  guix-build-46b905163530/output/powerpc64le-linux-gnu/bitcoin-46b905163530-powerpc64le-linux-gnu-debug.tar.gz
9e77d10a71ec9651f75979e1e86a4941ff65ac919e488ab97f0df6b9d8659e38  guix-build-46b905163530/output/powerpc64le-linux-gnu/bitcoin-46b905163530-powerpc64le-linux-gnu.tar.gz
222774b0f9c27f809f2a119060b6cdd5599d21ab2d155571b96ab268ab2e204f  guix-build-46b905163530/output/riscv64-linux-gnu/SHA256SUMS.part
974b2b931e1f1fb11280ade86f07a6ca9bf342a045888cfbe7c0bbc4f3f7a9ca  guix-build-46b905163530/output/riscv64-linux-gnu/bitcoin-46b905163530-riscv64-linux-gnu-debug.tar.gz
dcf447792f21adf7e7a89575431435f8ead62af7711eb4ef58ccde462afc4a81  guix-build-46b905163530/output/riscv64-linux-gnu/bitcoin-46b905163530-riscv64-linux-gnu.tar.gz
92b9bf892c21b78888f63e57a861bbe1da46d1d7acb818b59e0b428fbba3889c  guix-build-46b905163530/output/x86_64-apple-darwin/SHA256SUMS.part
cbdd177a07b36594275599b9c007eac647ad0c867deaca3730bc92b5b1feb3d7  guix-build-46b905163530/output/x86_64-apple-darwin/bitcoin-46b905163530-osx-unsigned.dmg
f065ff5a113e891d94744dfd8ee4f40889121d827730b85fd3725bee899f69a6  guix-build-46b905163530/output/x86_64-apple-darwin/bitcoin-46b905163530-osx-unsigned.tar.gz
10683266ada9544a9e59255dea32d476f655c9135bcc51d612af1eedc19c1b1e  guix-build-46b905163530/output/x86_64-apple-darwin/bitcoin-46b905163530-osx64.tar.gz
b9c1f758f15f68f18751cfa456f18e17d8cde5104137bb0a7ed9f7b048916b3a  guix-build-46b905163530/output/x86_64-linux-gnu/SHA256SUMS.part
02a341392452f0fe46fecfef09bfab9d8a99c4579568bca15480e3ae3676e6e3  guix-build-46b905163530/output/x86_64-linux-gnu/bitcoin-46b905163530-x86_64-linux-gnu-debug.tar.gz
6d9d6cb535694cd16285746a8fbafe2bd4752f4958ab47a36d2e1f6c958a3d23  guix-build-46b905163530/output/x86_64-linux-gnu/bitcoin-46b905163530-x86_64-linux-gnu.tar.gz
```

hebasto reviewed (approved):
re-ACK 5d399f9f3df513a0400049238f5ef0ef2352d57e

I've additionally verified building of `test_bitcoin.exe` with MSVC on Windows 10.

hebasto commented:
#### Guix builds:
```
$ find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
342270b55666026b8bb15ab628bdf13474358dbab3413360944c1dd63544ddd8  guix-build-5d399f9f3df5/output/aarch64-linux-gnu/SHA256SUMS.part
8d17415e272608b59e4f05b6632c66988f3ace386e2672ee6bbd6c208111f919  guix-build-5d399f9f3df5/output/aarch64-linux-gnu/bitcoin-5d399f9f3df5-aarch64-linux-gnu-debug.tar.gz
8f2e2049e18be8106024bed335be07afa2f1502097ae9dc8cb0d12e638add212  guix-build-5d399f9f3df5/output/aarch64-linux-gnu/bitcoin-5d399f9f3df5-aarch64-linux-gnu.tar.gz
c2df391b62efc9f43c7a81555636a1d6badbcf51e562e8dfa5c3a8b001892127  guix-build-5d399f9f3df5/output/arm-linux-gnueabihf/SHA256SUMS.part
013515a3e6ead9284eadbdb6b6d8167928cc4ca852f66a4f7e828f9b4f6b1997  guix-build-5d399f9f3df5/output/arm-linux-gnueabihf/bitcoin-5d399f9f3df5-arm-linux-gnueabihf-debug.tar.gz
5b036d03399b84dd6e57ea51cfa7007bdeaef6c89282ac99c2920dd2ca609091  guix-build-5d399f9f3df5/output/arm-linux-gnueabihf/bitcoin-5d399f9f3df5-arm-linux-gnueabihf.tar.gz
17cb8e6c004e92ca387c20c29798f698281257993a20aba43ddd7533ec52b35c  guix-build-5d399f9f3df5/output/arm64-apple-darwin/SHA256SUMS.part
151911e4488721c70068905c248256a16ff87ae4cb419c810ddb62a8b27f5d2d  guix-build-5d399f9f3df5/output/arm64-apple-darwin/bitcoin-5d399f9f3df5-arm64-apple-darwin.tar.gz
aae5f0448cf51ed340d7d13c59345c77cebea70d03ed43fb063a9cdb9f48944d  guix-build-5d399f9f3df5/output/arm64-apple-darwin/bitcoin-5d399f9f3df5-osx-unsigned.dmg
0db04585726aba2a1bcfe1ba0150a606d1f62c74da7ad792ff8bdf06d8a35dcd  guix-build-5d399f9f3df5/output/arm64-apple-darwin/bitcoin-5d399f9f3df5-osx-unsigned.tar.gz
eabb3b51338cafff4c3ae59bde6bb5990dc51c4a18e64affc0d51f3ba5dbde31  guix-build-5d399f9f3df5/output/dist-archive/bitcoin-5d399f9f3df5.tar.gz
d333668cfeae8f12640125b3e59282006cd448732b557eb16c6215289bdd2254  guix-build-5d399f9f3df5/output/powerpc64-linux-gnu/SHA256SUMS.part
1d4e7f94d2ba86296d08cf013800c70380a607bd4c036fe6aec7c2a9e086e419  guix-build-5d399f9f3df5/output/powerpc64-linux-gnu/bitcoin-5d399f9f3df5-powerpc64-linux-gnu-debug.tar.gz
3c311d7be2e99bc0525458a3d1e62981a0931e74fbc1b7fbc57bb6605c0467c3  guix-build-5d399f9f3df5/output/powerpc64-linux-gnu/bitcoin-5d399f9f3df5-powerpc64-linux-gnu.tar.gz
54236dd3475e8ec25bb4846deeb1559bd59e066ae5ca281ec1303c542c29b998  guix-build-5d399f9f3df5/output/powerpc64le-linux-gnu/SHA256SUMS.part
84d54e433ac333fc4375248222b505deddff27b2dc674e7d98025a23e85dd04a  guix-build-5d399f9f3df5/output/powerpc64le-linux-gnu/bitcoin-5d399f9f3df5-powerpc64le-linux-gnu-debug.tar.gz
015328a4dbf8dc4de81b481849adfbf3aa10c507822ef8a08c7b62bd31c95c57  guix-build-5d399f9f3df5/output/powerpc64le-linux-gnu/bitcoin-5d399f9f3df5-powerpc64le-linux-gnu.tar.gz
547692ca0c98f75dba521030c2dfba51264502ffd8469d382000baf2f1bc711d  guix-build-5d399f9f3df5/output/riscv64-linux-gnu/SHA256SUMS.part
c3ad252075a822a2141806cf6635a140d8790df02c5b7a045f824609b94c82c2  guix-build-5d399f9f3df5/output/riscv64-linux-gnu/bitcoin-5d399f9f3df5-riscv64-linux-gnu-debug.tar.gz
ae65d7c75d5881f01622a978c4c490b835a79ce0cbb224da6d56633432eaf635  guix-build-5d399f9f3df5/output/riscv64-linux-gnu/bitcoin-5d399f9f3df5-riscv64-linux-gnu.tar.gz
9ad380cf4e2e5aec5ede5dd4d0a27bc8de79d2f16a97e7a6bde441a164a7e9d8  guix-build-5d399f9f3df5/output/x86_64-apple-darwin/SHA256SUMS.part
9fb60025359b785e3c7577c2f6820e40dfb895eb6491fea2795748ec93004f1c  guix-build-5d399f9f3df5/output/x86_64-apple-darwin/bitcoin-5d399f9f3df5-osx-unsigned.dmg
dd2ef7250879cea902038e2967478a761080b66de4bb1cc87a1521696c0fd26f  guix-build-5d399f9f3df5/output/x86_64-apple-darwin/bitcoin-5d399f9f3df5-osx-unsigned.tar.gz
145c8cc6e739b6891d4611bcdcb5c58cffcebf9bc1689529a6b1f3878bcee7ec  guix-build-5d399f9f3df5/output/x86_64-apple-darwin/bitcoin-5d399f9f3df5-osx64.tar.gz
1a81f1365dc552b57815ee48150f76acff648505af73fcc7961b24dd370c1065  guix-build-5d399f9f3df5/output/x86_64-linux-gnu/SHA256SUMS.part
c462ab2acfd2485d526f8c08774d9384c746d6d6446d38783b93a46aafe3f17d  guix-build-5d399f9f3df5/output/x86_64-linux-gnu/bitcoin-5d399f9f3df5-x86_64-linux-gnu-debug.tar.gz
1c1b24233ad189157259258f263bf5c0baedb03b9de28f9f647c28eeb1425c51  guix-build-5d399f9f3df5/output/x86_64-linux-gnu/bitcoin-5d399f9f3df5-x86_64-linux-gnu.tar.gz
```

maflcko reviewed (approved):
Did not test on OpenBSD.

approach ACK 5d399f9f3df513a0400049238f5ef0ef2352d57e 📞

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

approach ACK 5d399f9f3df513a0400049238f5ef0ef2352d57e 📞
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUh1LgwAhNYbcUG8JabIlrScCLDnnJu9Fo+h6UoF/7uxjLDj5f0Iat87cLjeBgrP
ZEEXTmscefGzfA0sNblni7B7XIbMtJcLLE+RojMUKeKYVvE+9bwEyUMPwNGHCKnU
Z7yK0d2BbLAiZgkJdDab1f2Xjg0ZmrQSRl+GTRd2xkCrRRBACS9C7IV1eveGOBre
pfqHlbAo113U3V1at6nU/Q0YiG23teI/g8wLv9caeb5vNph6atwFkmLkHRwXOvO4
R94LccJ7Kh0xV0fqgas79ffy1Bpz9SEXuO1LX3y3cUOZqY5TVKLtaqrov/mjZsYE
w26/Z86s02WyOYWjGi/RUa3xigOwia7rGapWWvIH+U4LlB+ovlZq3PTwM4kucrtN
Ha5uNFK6CVMb4XTad8nyCkkDBPvY/9bS6Dqo975z1lPoCHnPKe4bKBSxIA94XMdH
3TAwkeUZmH7xzYuXiIkdimNygf1kl+oQtZJtf2xevfjQAySUzywxh89DTP7DxiEF
+Ruxzptk
=yjNN
-----END PGP SIGNATURE-----
```


</details>

MarcoFalke commented:
Tested `make check` on Bionic with system boost. Needs #24314 + https://github.com/bitcoin/bitcoin/pull/24254  or signer disabled.
